### PR TITLE
DDF-4133 Solr auth updates

### PR DESCRIPTION
**Pull request by ahoffer**
#### What does this PR do?
* Fixed formatting errors in documentation
* Improved logging and auditing
* Incorporated changes from Austin that got lost

#### Who is reviewing it? 
@austinsteffes 
@ricklarsen 
@paouelle 

#### Select relevant component teams: 
@codice/security 
@codice/solr 

#### Ask 2 committers to review/merge the PR and tag them here.
@clockard 
@vinamartin

#### How should this be tested?
This is how it was tested:
### TEST - DO NOT AUTOGENERATED PASSWORD
1. Set autogeneratePassword=false
2. Install
3. Log into Solr https://localhost:8994/solr
4. Username/password ddf/autogenerated 
5. Expected behavior
   * Solr login succeeeds
   * DF connects successfully to Solr cores
   * Alerts appear on Admin UI
   * Able to upload and search for a product
6. Actual behavior:
   *as expected
7. Test passed

### TEST - AUTOGENERATE PASSWORD
1. Set autogeneratedPassword=true
2. Restart
3. Expected behavior
   * Cannot log into Solr UI with ddf/autogenerated
   * DF connects successfully to Solr cores
   * Alerts appear on Admin UI
   * Able to upload and search for a product
   * Log messsages about password change exists
   * Security audit log messages about password change exists
4. Actual behavior
   * As expected
   * System log messages:
       * `2018-12-13T10:41:11,959 | INFO  | rClientAdapter 0 | SolrPasswordUpdateImpl           | .security.SolrPasswordUpdateImpl  166 | 374 - solr-password - 2.14.0.SNAPSHOT | "Updated encrypted Solr password in properties file /sandbox/ddf-2.14.0-SNAPSHOT/etc/custom.system.properties.`
       * `2018-12-13T10:41:11,960 | INFO  | rClientAdapter 0 | SolrPasswordUpdateImpl           | .security.SolrPasswordUpdateImpl  120 | 374 - solr-password - 2.14.0.SNAPSHOT | Updated encrypted Solr password in memory.`
       * `2018-12-13T10:41:11,942 | INFO  | rClientAdapter 0 | SolrPasswordUpdateImpl           | .security.SolrPasswordUpdateImpl  180 | 374 - solr-password - 2.14.0.SNAPSHOT | New password was set in Solr server.`
   * Security audit log messages:
       * `[INFO ] 2018-12-13T10:41:11,943 | rClientAdapter 0 | securityLogger  |  Subject: UNKNOWN New password was set in Solr server by the Solr password updater.`
       * `[INFO ] 2018-12-13T10:41:11,958 | rClientAdapter 0 | securityLogger  |  Subject: UNKNOWN "Updated encrypted Solr password in properties file /sandbox/ddf-2.14.0-SNAPSHOT/etc/custom.system.properties.`
6. Test passed
       
### TEST - MANUALLY RESTORE SOLR PASSWORD AND MANUALLY UPDATE SYSTEM PROPERTIES
1. Edit solr/server/solr/security.json and replace hashed password
2. Set autogeneratePassword to false in system properties file
3. Set solrPassword to "autogenerated" in system properties file
3. Start System
4. Expected behavior: DDF connects to Solr successfully
5. Actual behavior: as expected
7. Test passed

### NEGATIVE TEST - FAIL TO UPDATE SOLR PASSWORD
1. (From a clean distribution) Set start.solr to "false" in system properties file
5. Start System. DDF cannot communicate with Sorl because the Solr process was never started.
6. Expected behavior:
    * Solr password in system properties remains "autogenerated"
    * Message in log file that Solr password could not be updated
    * Message in audit log that Solr password could not be updated
7. Actual behavior: as expected
8. Test passed
    
### NEGATIVE TEST - FAIL TO UPDATE SYSTEM PROPERTIES FILE
1. (From a clean distribution) Set start.solr to "false" in system properties file
2. Put in a break point where DDF would save the new password in the system properties file
3. Start System with debugger attached.
4. When the break point is hit, remove all write permission from the custom.system.properites file. This will prevent DDF from updating the file.
3. Expected behavior:
   * Messages in log and audit file that Solr password was updated, but not saved.
   * DDF will not be able to communicate with Solr (because the Solr process was updated, but the new password was not saved so DDF does not know the new password)
4. Actual behavior:
    * `2018-12-13T15:19:45,081 | ERROR | rClientAdapter 0 | SolrPasswordUpdateImpl           | .security.SolrPasswordUpdateImpl  172 | 374 - solr-password - 2.14.0.SNAPSHOT | Exception while writing to /sandbox/ddf-2.14.0-SNAPSHOT/etc/custom.system.properties. Solr password was changed, but new password was not saved.`
    * `[INFO ] 2018-12-13T15:19:00,466 | rClientAdapter 0 | securityLogger  |  Subject: UNKNOWN New password was set in Solr server by the Solr password updater.`
5. Test passed

#### Any background context you want to provide?
No.

#### What are the relevant tickets?
[[[DDF-4133]](https://codice.atlassian.net/browse/DDF-4133)](https://codice.atlassian.net/browse/DDF-4133)

#### Checklist:
- [X] Documentation Updated
- [X] Update / Add Unit Tests
- [NA] Update / Add Integration Tests

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

**Comment by ahoffer** [comment]
build now

File: platform/security/encryption/platform-security-encryption-impl/src/main/java/ddf/security/encryption/impl/EncryptionServiceImpl.java [hunk 1]
```diff
@@ -117,9 +117,12 @@ public String decryptValue(String wrappedEncryptedValue) {
 
   @Override
   public String encryptValue(String unwrappedPlaintext) {
-    if (unwrappedPlaintext == null || unwrappedPlaintext.equals("")) {
+    if (unwrappedPlaintext == null) {
```
Comment (vinamartin) [inline platform/security/encryption/platform-security-encryption-impl/src/main/java/ddf/security/encryption/impl/EncryptionServiceImpl.java]: ❓ Can you do something like `if (unwrappedPlaintext.isEmpty()) return unwrappedPlaintext;` ?
http://commons.apache.org/proper/commons-lang/javadocs/api-2.6/org/apache/commons/lang/StringUtils.html#isEmpty(java.lang.String)

File: platform/solr/solr-password/src/main/java/ddf/platform/solr/security/SolrPasswordUpdateImpl.java [hunk 2]
```diff
@@ -161,13 +161,16 @@ private Void updatePropertiesFile() {
...
+              "Updated encrypted Solr password in properties file %s.", systemPropertyFilename);
       SecurityLogger.audit(msg);
       LOGGER.info(msg);
     } catch (IOException e) {
-      LOGGER.error(
-          "Exception while writing to {}. Solr password was changed, but new password was not saved.",
-          systemPropertyFilename);
+      String msgFail =
+          String.format(
+              "Exception while writing to %s. Solr password was changed, but new password was not saved.",
```
Comment (vinamartin) [inline platform/solr/solr-password/src/main/java/ddf/platform/solr/security/SolrPasswordUpdateImpl.java]: ❓ Is there any way to recover from this? Maybe we could also give some kind of action to take.

File: platform/security/encryption/platform-security-encryption-impl/src/main/java/ddf/security/encryption/impl/EncryptionServiceImpl.java (hunk 1 above)
Comment (austinsteffes) [reply platform/security/encryption/platform-security-encryption-impl/src/main/java/ddf/security/encryption/impl/EncryptionServiceImpl.java]: there needs to be two separate things returned if null or "" and isempty() would instead catch both of those. based on comment made on other PR https://github.com/codice/ddf/pull/4083#discussion_r239947550.

Comment (vinamartin) [reply platform/security/encryption/platform-security-encryption-impl/src/main/java/ddf/security/encryption/impl/EncryptionServiceImpl.java]: I'm confused it looks like you're returning whatever the value of unwrappedPlaintext is for both scenarios:

null -> return null
"" -> return ""

It won't work to return it?

Comment (austinsteffes) [reply platform/security/encryption/platform-security-encryption-impl/src/main/java/ddf/security/encryption/impl/EncryptionServiceImpl.java]: ah, sorry i was confused. yes you're correct good suggestion

File: platform/solr/solr-password/src/main/java/ddf/platform/solr/security/SolrPasswordUpdateImpl.java (hunk 2 above)
Comment (ahoffer) [reply platform/solr/solr-password/src/main/java/ddf/platform/solr/security/SolrPasswordUpdateImpl.java]: We could point them to the documentation. There are steps there to reset the password in Solr. But I do not know of anyone log messages where we say "go read the docs". Is there something else we say that would be useful? Maybe "the application cannot use Solr until the password is restored"? What do you think?

Comment (vinamartin) [reply platform/solr/solr-password/src/main/java/ddf/platform/solr/security/SolrPasswordUpdateImpl.java]: Yeah I like "the application cannot use solr until the password is restored" might be helpful and then maybe from there they'd figure out they need to consult the docs.

Comment (austinsteffes) [reply platform/solr/solr-password/src/main/java/ddf/platform/solr/security/SolrPasswordUpdateImpl.java]: this and others will give discrepancies between release and master - These are good suggestions and I agree with them just putting this comment out there for future reference

File: platform/security/encryption/platform-security-encryption-impl/src/main/java/ddf/security/encryption/impl/EncryptionServiceImpl.java (hunk 1 above)
Comment (ahoffer) [reply platform/security/encryption/platform-security-encryption-impl/src/main/java/ddf/security/encryption/impl/EncryptionServiceImpl.java]: I originally used StringUtils, but realized I was bringing in Apache Commons Lang just to check for an empty string. I can certainly return the input. I also update `decryptValue` to do the same.

File: platform/solr/solr-password/src/main/java/ddf/platform/solr/security/SolrPasswordUpdateImpl.java (hunk 2 above)
Comment (ahoffer) [reply platform/solr/solr-password/src/main/java/ddf/platform/solr/security/SolrPasswordUpdateImpl.java]: Updated message.

**Comment by ahoffer** [comment]
build now

File: platform/security/encryption/platform-security-encryption-impl/src/main/java/ddf/security/encryption/impl/EncryptionServiceImpl.java (hunk 1 above)
Comment (paouelle) [reply platform/security/encryption/platform-security-encryption-impl/src/main/java/ddf/security/encryption/impl/EncryptionServiceImpl.java]: So, why no not keep the test as it was before and return `unwrappedPlaintext`? Same in `decryptValue()` above?

**Comment by ahoffer** [comment]
build now

**Comment by ahoffer** [comment]
build now

File: platform/security/encryption/platform-security-encryption-impl/src/main/java/ddf/security/encryption/impl/EncryptionServiceImpl.java [hunk 3]
```diff
@@ -97,15 +98,12 @@ public synchronized String decrypt(String encryptedValue) {
...
-    String encryptedValue = unwrapEncryptedValue(wrappedEncryptedValue);
-    if (wrappedEncryptedValue == null) {
-      LOGGER.debug("A null password was provided.");
-      return null;
-    }
-    if (wrappedEncryptedValue.isEmpty()) {
-      LOGGER.debug("A blank password was provided in the configuration.");
-      return "";
+    if (StringUtils.isEmpty(wrappedEncryptedValue)) {
+      return wrappedEncryptedValue;
```
Comment (bakejeyner) [inline platform/security/encryption/platform-security-encryption-impl/src/main/java/ddf/security/encryption/impl/EncryptionServiceImpl.java]: ❓  Should we add debug logging here?

Comment (ahoffer) [reply platform/security/encryption/platform-security-encryption-impl/src/main/java/ddf/security/encryption/impl/EncryptionServiceImpl.java]: I thought about it and I didn't see the value. It the previous logging statements felt like noise to me.

**Comment by ahoffer** [comment]
network failure. build now.

**Pull request merged**